site/en/blog/blinkng/index.md

```diff
@@ -0,0 +1,185 @@
+
+-  Does the output of style, layout, or paint need to be updated?
+-  When will these data get their "final" value?
+-  When is it OK to modify these data?
+-  When will this object be deleted?
+
+There are many examples of this, including:
+
+_Style_ would generate ComputedStyle based on stylesheets; but ComputedStyle was not immutable; in some cases it would be modified by later pipeline stages.
+
+_Style_ would generate a tree of `LayoutObject`, and then _layout_ would annotate those objects with size and positioning information. In some cases, _layout_ would even modify the tree structure. There was no clear separation between the inputs and outputs of _layout_.
+
+_Style_ would generate accessory data structures that determined the course of _compositing_, and those data structures were modified in place by every phase after _style_.
+
+At a lower level, rendering data types largely consists of specialized trees (for example, the DOM tree, style tree, layout tree, paint property tree); and rendering updates are implemented as recursive tree walks. Ideally, a tree walk should be _contained_: when processing a given tree node, we should not access any information outside of the subtree rooted at that node. That was never true pre-RenderingNG; tree walks frequently accessed information from the ancestors of the node being processed. This made the system very fragile and error-prone. It was also impossible to begin a tree walk from anywhere but the root of the tree.
+
+Finally, there were many on-ramps into the rendering pipeline sprinkled throughout the code: forced layouts triggered by JavaScript, partial updates triggered during document load, forced updates to prepare for event targeting, scheduled updates requested by the display system, and specialized APIs exposed only to test code, to name a few. There were even a few _recursive_ and _reentrant_ paths into the rendering pipeline (that is jumping to the beginning of one stage from the middle of another). Each of these on-ramps had their own idiosyncratic behavior, and in some cases the output of rendering would depend on the manner in which the rendering update was triggered.
+
+## What we changed
+
+RenderingNG is composed of many sub-projects, big and small, all with the shared goal of eliminating the architectural deficits described earlier. These projects share a few guiding principles designed to make the rendering pipeline more of an actual pipeline:
+
+-  **Uniform point of entry**: We should always enter the pipeline at the beginning.
+-  **Functional stages**: Each stage should have well-defined inputs and outputs, and its behavior should be _functional_, that is, deterministic and repeatable, and the outputs should depend only on the defined inputs.
+-  **Constant inputs**: The inputs of any stage should be effectively constant while the stage is running.
+-  **Immutable outputs**: The output of any stage should be immutable once the stage is finished.
```

Comment:
Change the description to:

Once a stage has finished, its outputs should be immutable for the remainder of the rendering update.